SingularityService/src/main/java/com/hubspot/singularity/resources/TaskResource.java

```diff
@@ -656,4 +668,56 @@ public Response downloadFileOverProxy(
     }
 
   }
+
+  private static class NingOutputToJaxRsStreamingOutputWrapper implements AsyncHandler<Void>, StreamingOutput {
+    private OutputStream wrappedOutputStream;
+    private BoundRequestBuilder requestBuilder;
+
+    public NingOutputToJaxRsStreamingOutputWrapper(BoundRequestBuilder requestBuilder) {
+      this.requestBuilder = requestBuilder;
+    }
+
+    @Override
+    public void onThrowable(Throwable t) {
+      LOG.error("Unable to proxy file download", t);
+    }
+
+    @Override
+    public STATE onBodyPartReceived(HttpResponseBodyPart bodyPart) throws Exception {
+      bodyPart.writeTo(wrappedOutputStream);
+      wrappedOutputStream.flush();
+      return STATE.CONTINUE;
+    }
+
+    @Override
+    public STATE onStatusReceived(HttpResponseStatus responseStatus) {
+      LOG.trace("Proxying download of {} from Mesos: Status={}", requestBuilder.build().getUrl(), responseStatus.getStatusCode());
+      return STATE.CONTINUE;
+    }
+
+    @Override
+    public STATE onHeadersReceived(HttpResponseHeaders headers) {
+      LOG.trace("Proxying download of {} from Mesos: Headers={}", requestBuilder.build().getUrl(), headers.getHeaders());
+      return STATE.CONTINUE;
+    }
+
+    @Override
+    public Void onCompleted() {
+      LOG.info("Proxying download of {} from Mesos: Completed.", requestBuilder.build().getUrl());
+      return null;
+    }
+
+    // StreamingOutput implementation: just make the OutputStream available to the AsyncHandler<> implementation.
+    @Override
+    public void write(OutputStream output) throws WebApplicationException, IOException {
+      if (wrappedOutputStream == null) {
+        wrappedOutputStream = output;
+        try {
+          requestBuilder.execute(this).get();
```

Comment:
We need to do it this way because the `OutputStream` doesn't become available to us until Jersey actually calls `write()` on this implementation. The alternative would be to begin the proxied request _before_ waiting for Jersey, in which case we'd be back to buffering a potentially large response in memory.